File: EnergyAPI/src/main/java/fr/traqueur/energylib/api/persistents/EnergyTypePersistentDataType.java
package fr.traqueur.energylib.api.persistents;

import fr.traqueur.energylib.api.types.EnergyType;
import org.bukkit.persistence.PersistentDataAdapterContext;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;

/**
 * This class is used to save the EnergyType in a ItemStack
 */
public class EnergyTypePersistentDataType implements PersistentDataType<String, EnergyType> {

    /**
     * The instance of the class
     */
    public static final EnergyTypePersistentDataType INSTANCE = new EnergyTypePersistentDataType();

    /**
     * Get the primitive type of the class
     * @return the primitive type
     */
    @NotNull
    @Override
    public Class<String> getPrimitiveType() {
        return String.class;
    }

    /**
     * Get the complex type of the class
     * @return the complex type
     */
    @NotNull
    @Override
    public Class<EnergyType> getComplexType() {
        return EnergyType.class;
    }

    /**
     * Convert the EnergyType to a primitive type
     * @param energyType the EnergyType
     * @param persistentDataAdapterContext the context
     * @return the primitive type
     */
    @NotNull
    @Override
    public String toPrimitive(@NotNull EnergyType energyType, @NotNull PersistentDataAdapterContext persistentDataAdapterContext) {
        return energyType.getName();
    }

    /**
     * Convert the primitive type to a EnergyType
     * @param s the primitive type
     * @param persistentDataAdapterContext the context
     * @return the EnergyType
     */
    @NotNull
    @Override
    public EnergyType fromPrimitive(@NotNull String s, @NotNull PersistentDataAdapterContext persistentDataAdapterContext) {
        return EnergyType.TYPES.stream()
                .filter(type -> type.getName().equals(s))
                .findFirst()
                .orElseThrow();
    }
}

File: FluxAPI/src/main/kotlin/io/github/Jadiefication/redstoneflux/api/persistents/EnergyTypePersistentDataType.kt
package io.github.Jadiefication.redstoneflux.api.persistents

import io.github.Jadiefication.redstoneflux.api.types.EnergyType
import org.bukkit.persistence.PersistentDataAdapterContext
import org.bukkit.persistence.PersistentDataType

/**
 * This class is used to save the EnergyType in a ItemStack
 */
class EnergyTypePersistentDataType : PersistentDataType<String, EnergyType> {
    /**
     * Get the primitive type of the class
     * @return the primitive type
     */
    override fun getPrimitiveType(): Class<String?> {
        return String::class.java as Class<String?>
    }

    /**
     * Get the complex type of the class
     * @return the complex type
     */
    override fun getComplexType(): Class<EnergyType?> {
        return EnergyType::class.java as Class<EnergyType?>
    }

    /**
     * Convert the EnergyType to a primitive type
     * @param energyType the EnergyType
     * @param persistentDataAdapterContext the context
     * @return the primitive type
     */
    override fun toPrimitive(
        energyType: EnergyType,
        persistentDataAdapterContext: PersistentDataAdapterContext
    ): String {
        return energyType.name!!
    }

    /**
     * Convert the primitive type to a EnergyType
     * @param s the primitive type
     * @param persistentDataAdapterContext the context
     * @return the EnergyType
     */
    override fun fromPrimitive(s: String, persistentDataAdapterContext: PersistentDataAdapterContext): EnergyType {
        return EnergyType.Companion.TYPES.stream()
            .filter { type: EnergyType? -> type!!.name == s }
            .findFirst()
            .orElseThrow()!!
    }

    companion object {
        /**
         * The instance of the class
         */
        val INSTANCE: EnergyTypePersistentDataType = EnergyTypePersistentDataType()
    }
}
